src/lib/SaferNames/Bridge.hs

```diff
@@ -0,0 +1,220 @@
+-- Copyright 2019 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module SaferNames.Bridge (HasSafeVersionE (..), HasSafeVersionB (..))  where
+
+import Data.Foldable (toList)
+import qualified Data.Set as Set
+
+import LabeledItems
+import Syntax
+import Env
+
+import SaferNames.Name
+import SaferNames.Syntax
+
+import qualified Syntax as D  -- D for Danger
+import qualified Env    as D
+
+import qualified SaferNames.Syntax as S
+import qualified SaferNames.Name   as S
+
+class HasSafeVersionE (e:: *) where
+  type SafeVersionE e :: S.E
+  toSafeE   :: e -> SafeVersionE e n
+  fromSafeE :: SafeVersionE e n -> e
+
+class HasSafeVersionB (b:: *) where
+  type SafeVersionB b :: S.B
+  toSafeB   :: b -> SafeVersionB b n l
+  fromSafeB :: SafeVersionB b n l -> b
+
+instance HasSafeVersionE D.Name where
+  type SafeVersionE D.Name = S.Name
+  toSafeE name = UnsafeMakeName name
+  fromSafeE (UnsafeMakeName name) = name
+
+instance HasSafeVersionE D.Atom where
+  type SafeVersionE D.Atom = S.Atom
+
+  toSafeE atom = case atom of
+    D.Var v -> S.Var $ toSafeE v
+    D.Lam (D.Abs b (arr, body)) ->
+      S.Lam $ S.Abs (toSafeB b) (S.WithArrow (fmap toSafeE arr) (toSafeE body))
```

Comment:
Do places like those assume that the dangerous representation already has no shadowing, or is that fine because this is not `FreshAbs`?

Reply:
That's right. Ordinary `Abs` doesn't have a freshness guarantee. (`FreshAbs` explicitly carries a `FreshExt`, which is the only way you get a freshness guarantee.)